Proper handling of arguments passed by in kwargs inside zip_schema

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137311

if the function is

```func(a, b, c) ```
and is called as 
```func(a=1, b=.., c=..)```
before this change we do not iterate on the a, b, c, since those appear in kwargs this diff fix that issue.

This function is used in _inductor/ir.py to iterate over custom op arguments and when a custom pass does changes 
and pass arguments as kwargs, we do not process them. 
```
        for info, arg in torch._library.utils.zip_schema(schema, args, kwargs):
            handle_aliasing_and_mutation(info, arg)
```
Fix https://github.com/pytorch/pytorch/issues/137057

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

laithsakka committed:
Proper habdle of args in kwargs in zip_schema

[ghstack-poisoned]

laithsakka committed:
Update on "Proper habdle of args in kwargs in zip_schema"




[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

zou3519 commented on torch/_library/utils.py:
```diff
@@ -193,6 +193,8 @@ def zip_schema(
                 yield info, kwargs[info.name]
             continue
         if i >= len(args):
+            if not info.kwarg_only and info.name in kwargs:
+                yield info, kwargs[info.name]
```
@laithsakka can you add a test?

zou3519 commented on torch/_library/utils.py:
```diff
@@ -193,6 +193,8 @@ def zip_schema(
                 yield info, kwargs[info.name]
             continue
         if i >= len(args):
+            if not info.kwarg_only and info.name in kwargs:
```
@laithsakka the docstring for lines 185 to 186 need to be updated after this change. I'm not really sure what the right way to put this is, maybe "(args, kwargs) must be bindable to the schema (args, kwargs)"

zou3519 reviewed (approved):
please add a test and update the docstring. Otherwise, this is awesome -- great find

laithsakka committed:
Update on "Proper handling of arguments passed by in kwargs inside zip_schema"


if the function is

```func(a, b, c) ```
and is called as 
```func(a=1, b=.., c=..)```
before this change we do not iterate on the a, b, c, since those appear in kwargs this diff fix that issue.

This function is used in _inductor/ir.py to iterate over custom op arguments and when a custom pass does changes 
and pass arguments as kwargs, we do not process them. 
```
        for info, arg in torch._library.utils.zip_schema(schema, args, kwargs):
            handle_aliasing_and_mutation(info, arg)
```
Fix https://github.com/pytorch/pytorch/issues/137057

[ghstack-poisoned]

laithsakka commented:
add unit test and update comment

laithsakka commented:
@pytorchbot merge
